### DISCUSSION TITLE
Is there a limit to the Goroutine that calls gen.ServerProcess.CallWithTimeout (...)?

### DISCUSSION THREAD

**Discussion by leonlee2013** [ergo-services/ergo#132] [Q&A] [answered]
`CallWithTimeout (to, message) `can only be called in Goroutine for running ServerBehavior？

```go
HandleCast(process *ServerProcess, message etf.Term) ServerStatus
HandleCall(process *ServerProcess, from ServerFrom, message etf.Term) (etf.Term, ServerStatus)
HandleDirect(process *ServerProcess, ref etf.Ref, message interface{}) (interface{}, DirectStatus)
HandleInfo(process *ServerProcess, message etf.Term) ServerStatus
 ```
Otherwise, ProcessLoop will be blocked ： ` case sp.waitReply = <-sp.callbackWaitReply:`
```go
func (sp *ServerProcess) waitCallbackOrDeferr(message interface{}) {
 ...
  select {

  //case <-sp.Context().Done():
  // do not read the context state. otherwise the goroutine with running callback
  // might lock forever on exit (or on making a Call request) as nobody read
  // the callbackWaitReply channel.

  case sp.waitReply = <-sp.callbackWaitReply:
    // not nil value means callback made a Call request and waiting for reply
    if sp.waitReply == nil && len(sp.deferred) > 0 { 
      sp.mailbox = sp.deferred
    }   
    return
  }
}
```

**Comment by halturin** [comment, marked as answer]
Call*/Cast methods can be used inside gen.ServerProcess only (within Handle* callbacks only, to be specific). If you need to make a sync call to the process being outside of the actor, you should use gen.Process.Direct or gen.Process.DirectWithTimeout or gen.Process.Send for async messaging

**Comment by halturin** [comment]
btw, here is auto-translation of the documentation for the gen.Process https://ru-docs-ergo-services.translate.goog/basics/process?_x_tr_sl=auto&_x_tr_tl=en&_x_tr_hl=en&_x_tr_pto=wapp. Its still a draft, but hope it will help you to understand some details.

**Reply by leonlee2013** [reply]
Thanks for your reply!
Calls like the one below are also not allowed, right?
```go
  func (demo *Demo) HandleCall(process *gen.ServerProcess, from gen.ServerFrom, message etf.Term) (etf.Term, gen.ServerStatus) {
    ...
    go func() {
     ret, err := eh.ServerProcess.CallRPCWithTimeout(1, node, module, f, args...)
      if err != nil {
        panic(err)
      }
    }()
    return message, gen.ServerStatusOK
  }
```

**Reply by halturin** [reply]
not sure what the `eh.ServerProcess` is, but it definitely the wrong way of usage.

**Reply by leonlee2013** [reply]
Sorry, I wrote it wrong. It should  like this：
```go
  func (demo *Demo) HandleCall(process *gen.ServerProcess, from gen.ServerFrom, message etf.Term) (etf.Term, gen.ServerStatus) {
    ...
    go func() {
     ret, err := process.CallWithTimeout(to, message, timeout )
      if err != nil {
        panic(err)
      }
    }()
    return message, gen.ServerStatusOK
  }
```
Called  `process.CallWithTimeout(to, message, timeout )` in the go func(){}()

**Reply by halturin** [reply]
As mentioned above, you should use Call/Cast within Handle* callback only. Using them outside of the `process` goroutine causing a block. As for your example, starting a new goroutine makes this messaging async, so I recommend using Send or Cast methods.

**Reply by leonlee2013** [reply]
Okay, thanks

**Reply by leonlee2013** [reply]
> As mentioned above, you should use Call/Cast within Handle* callback only. Using them outside of the process goroutine causing a block. 

`gen.ServerProcess.Cast(...)` is asynchronous and should not require,
`gen.ServerProcess.Call(...)` must be within Handle* callback only. 
```go
func (sp *ServerProcess) Cast(to interface{}, message etf.Term) error {
  msg := etf.Term(etf.Tuple{etf.Atom("$gen_cast"), message})
  return sp.Send(to, msg)
}
func (sp *ServerProcess) Call(to interface{}, message etf.Term) (etf.Term, error) {
  return sp.CallWithTimeout(to, message, DefaultCallTimeout)
}
func (sp *ServerProcess) CallWithTimeout(to interface{}, message etf.Term, timeout int) (etf.Term, error) {
  ref := sp.MakeRef()
  from := etf.Tuple{sp.Self(), ref}
  msg := etf.Term(etf.Tuple{etf.Atom("$gen_call"), from, message})

  sp.PutSyncRequest(ref)
  if err := sp.Send(to, msg); err != nil {
    sp.CancelSyncRequest(ref)
    return nil, err
  }
  sp.callbackWaitReply <- &ref
  value, err := sp.WaitSyncReply(ref, timeout)
  return value, err
}
```

**Reply by halturin** [reply]
sure, the Cast method is just a wrap over the gen.Process.Send method. But it is placed along with the Call method because it handles by the HandleCast callback. If you decide to implement a custom Process behavior (like your own ServerProcess version) you should also make sure to implement the handling of `etf.Term(etf.Tuple{etf.Atom("$gen_cast"), message})`

**Reply by leonlee2013** [reply]
`node.process.Direct(...)` for `gen.Application,gen.supervisor,gen.Server`
`gen.ServerProcess.Cast(...)`  is only used by `gen.server`, 
so there is no `node.process.Cast(...)` 
Is my understanding correct?

**Reply by halturin** [reply]
requests made with `gen.Process.Direct(...)` are handled by `gen.Application`, `gen.Supervisor` and `gen.Server`, but only the last one (`gen.Server`) forwards this request to the HandleDirect callback. (`gen.Application` and `gen.Supervisor` are handling this kind of message under the hood)

> gen.ServerProcess.Cast(...) is only used by gen.server, so there is no node.process.Cast(...)

that's correct
